# Analyzing Used Car Listings - Exploring German eBay Car Webiste 

A [classified](https://en.wikipedia.org/wiki/Classified_advertising) advertisement is a form of advertising, commonly found in newspapers, online or other periodicals, which is sold or distributed free of charge. 

**eBay Kleinanzeigen** is a classified section of **German** eBay website. 

![German eBay Website](http://localhost:8888/files/Desktop/DQ_Projects/Gebrauchtwagen-eBay-Kleinanzeigen.png "German eBay Website")

We will be mainly working with the used cars classifieds section of the eBay website. The [dataset](https://data.world/data-society/used-cars-data) consists information of used cars from the website. 

The aim of the project is to analyze the included used car listings and answer few questions: <br>

**1. What is the average price of the top 6 brands in the used car listings?** <br>
**2. Is there a link between the top 6 brands' car price and the average mileage?**

The *data dictionary* for the dataset is below:

| Field Name           | Description                                                                                       |
|----------------------|----------------------------------------------------------------------------------------------------|
| dateCrawled          | The date when the ad was first crawled. All field-values are taken from this date.                 |
| name                 | The name of the car.                                                                               |
| seller               | Whether the seller is a private individual or a dealer.                                           |
| offerType            | The type of listing (for sale or exchange).                                                        |
| price                | The price of the car in the ad.                                                                    |
| abtest               | Whether the listing is included in an A/B test.                                                   |
| vehicleType          | The type of vehicle (e.g. SUV, sedan, coupe, etc.).                                                |
| yearOfRegistration   | The year in which the car was first registered.                                                   |
| gearbox              | The type of transmission (e.g. manual, automatic, etc.).                                           |
| powerPS              | The power of the car in PS (Pferdestärke, a German unit of power, roughly equal to horsepower).     |
| model                | The name of the car model.                                                                         |
| odometer             | The number of kilometers the car has been driven.                                                  |
| monthOfRegistration  | The month in which the car was first registered.                                                   |
| fuelType             | The type of fuel the car uses (e.g. gasoline, diesel, electric, etc.).                             |
| brand                | The brand of the car (e.g. Toyota, Ford, BMW, etc.).                                               |
| notRepairedDamage    | Whether the car has any unrepaired damage.                                                         |
| dateCreated          | The date on which the eBay listing was created.                                                    |
| nrOfPictures         | The number of pictures in the ad.                                                                  |
| postalCode           | The postal code for the location of the vehicle.                                                   |
| lastSeenOnline       | The date when the crawler last saw this ad online.                                                 |

Note: PS (Pferdestärke) is a German unit of power, which is roughly equal to horsepower (hp).

In [1]:
# future warnings to be ignored
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

# importing libraries
import numpy as np
import pandas as pd

In [2]:
# reading the csv file into a dataframe
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

# Exploring the dataset

In [4]:
# exploring infomation about the dataset and a few rows
autos.info()
print('\n')
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### A few observations of the dataset

* There are `50,000 rows` and `20 columns` in the dataset with various information about the cars like seller, the year it was registered, model, etc.
* The data-types in the dataset are of various kinds - object(string) and integers.
* Some columns have null values in them, but none have more than ~20% null values.
* The column names use [camel case](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snake case](https://en.wikipedia.org/wiki/Snake_case) which means we can't just replace space with underscores.

## Changing the column names

Like we mentioned above, the column names are in camel case and we would like to change it to snake case and make the column names a bit more descriptive based on the information in the columns. Having descriptive column names ensures the we have more details about the type of data we are working with while performing analysis.

Firstly, let's have a look at all the column names

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# changing column names to snake case and making it more descriptive
# new column names
new_column_names = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
                   'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
                   'odometer_km', 'registration_month', 'fuel_type', 'brand',
                   'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code', 'last_seen']

# assigning the new column names 
autos.columns = new_column_names

In [7]:
# checking the new column names 
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   

   power_ps   model odometer_km  registration_month fuel_type    brand  \
0       158  andere   150,000km                   3       lpg  peugeot   
1       286     7er   150,000km                   6    benzin      bmw   

  unrepaired_damage           ad_created  no_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08


We converted all the column names from camel case to snake case and reworded a few column names to make it more descriptive based on the data dictionary.

## Cleaning the data

Now that we have converted all the column names according to the need of the analysis. We will have to clean the dataset to make it uniform and have only the columns which will add value to the analysis.

Initially, we will look for:
1. **Text columns where all or almost all values are the same.** Why? Because they don't have useful information for analysis, so these can often be dropped.
2. **Numeric data which are stored as text.** These needs to be cleaned and converted. Why? Since, they are stored as strings, we won't be able to perform numeric calculations on them. So, they need to be coverted to an integer.

In [8]:
# descriptive statistics for all columns
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

       odometer_km  registration_month fuel_type       brand  \
count        50000        50000.000000     45518       50000   
unique          13                 NaN         7          40   
top      150,000km                 NaN    benzin  volkswagen   
freq         32424                 NaN     30107       10687   
mean           NaN            5.723360       NaN         NaN   
std            NaN            3.711984       NaN         NaN   
min            NaN            0.000000       NaN         NaN   
25%            NaN            3.000000       NaN         NaN   
50%            NaN            6.000000       NaN         NaN   
75%            NaN            9.000000       NaN         NaN   
max            NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  no_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Text columns where all or almost all values are the same.

We can see that the values in the columns - **`seller`, `offer_type`, `abtest`, `gear_box`, `unrepaired_damage`** - have all or almost the same values. <br>

Let's have a closer look at these columns and drop some of the columns which won't be of much weightage to our analysis. 

In [9]:
for column in ['seller', 'offer_type', 'abtest', 'gearbox', 'unrepaired_damage']:
    print(autos[column].value_counts())
    print('-' * 30)

privat        49999
gewerblich        1
Name: seller, dtype: int64
------------------------------
Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64
------------------------------
test       25756
control    24244
Name: abtest, dtype: int64
------------------------------
manuell      36993
automatik    10327
Name: gearbox, dtype: int64
------------------------------
nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64
------------------------------


Looking at the above value counts of text columns which have almost same values, we can see that, for the **`seller`** and **`offer_type`** columns, the count of values are majorly towards a single value. Hence, we will drop these columns.

In [10]:
# dropping the columns mentioned above

print(autos.shape) # shape before dropping the columns
autos = autos.drop(columns=['seller', 'offer_type'])
print(autos.shape) # shape after dropping the columns

(50000, 20)
(50000, 18)


We have dropped the columns mentioned above. Before moving on to the next step, let's have a closer look at another column -**`no_of_pictures`**. As it looks a bit of error prone.

In [11]:
autos['no_of_pictures'].value_counts()

0    50000
Name: no_of_pictures, dtype: int64

We can see that all the rows of the listing contains no pictures at all. This won't add much purpose to the analysis. So, we will get rid of this column, as well.

In [12]:
autos = autos.drop(columns='no_of_pictures')

In [13]:
# checking all the columns after dropping 
autos.head(3)

date_crawled                                        name   price  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  $5,000   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  $8,500   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  $8,990   

    abtest vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004    manuell       158  andere   
1  control    limousine               1997  automatik       286     7er   
2     test    limousine               2009    manuell       102    golf   

  odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0   150,000km                   3       lpg     peugeot              nein   
1   150,000km                   6    benzin         bmw              nein   
2    70,000km                   7    benzin  volkswagen              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00        35394  2016-04-06 20:15:37

### Numeric data which are stored as text.

Now that we have gotten rid of the columns which didn't add any value, there are few columns which are **numbers but are stored as a string.** We need to convert them, so that we will be able to perform numeric calculations on these columns, if required.

In [14]:
# columns which are stored as strings
for column in ['price', 'odometer_km']:
    print(autos[column].describe())
    print('-' * 30)

count     50000
unique     2357
top          $0
freq       1421
Name: price, dtype: object
------------------------------
count         50000
unique           13
top       150,000km
freq          32424
Name: odometer_km, dtype: object
------------------------------


We can see that the above two columns, **`price` and `odometer_km`** which are actually supposed to be of `int` type are of the `object` type, which are strings. We need to convert them to numeric data. <br>

We can see that the `price` column has a **$ (Dollar)** symbol and **comma** in `odometer_km` column, so we need to remove that before converting to an integer. 

In [15]:
# converting the price column
autos['price'] = (autos['price']
                .str.replace('$', '')
                .str.replace(',', '')
                .astype(int))
            
# converting the odometer_km column
autos['odometer_km'] = (autos['odometer_km']
                        .str.replace('km', '')
                        .str.replace(',', '')
                        .astype(int))

Now we have converted the above two columns to the integer data type. Let's describe the columns to check their data type and see a few summarized statistics of these columns.

In [16]:
for column in ['price', 'odometer_km']:
    print(autos[column].describe())
    print('-' * 30)

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
------------------------------
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
------------------------------


### Finding outliers

We will continue exploring the data looking for values which are unrealistically high or low **(outliers)**. We will look for them in the numeric columns of the `price` and `odometer_km` columns.

We will first look at the `odometer_km` column for any outliers.

In [17]:
# exploring the odometer_km column
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

* We can see that the **75th percentile and maximum** odometer value of cars are the same i.e **150,000 km**.
* So, the 150,000 is not really an outlier because majority of the cars mileage are in the range of 125,000 km - 150,000 km

Let's explore the `odometer_km` column in depth

In [18]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

* The minimum mileage of the car listed are **5000 km** and not **0**, this makes sense because the cars mentioned on the site are used cars.
* Most of the cars listed have mileage more than 50,000 km.

Now let's take a look at the `price` column for any outliers.

In [19]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [20]:
autos['price'].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

* The minimum price of a car is **0 USD**, which doesn't make sense at all. All the car listed are for sale, so it should have a price. So, the cars with price 0 are outliers and it will affect our analysis.
* The maximum price is **100,000,000 USD** and 75th percentile is **7,200 USD**. So, the price of 100,000,000 USD is not justified. So, that's an outlier, as well.
* There are 1421 rows with the price 0 USD.

So, we will keep only the rows which have a price which is **greater than 0 USD and less than 100,000,000 USD.**

In [21]:
autos = autos[autos['price'].between(1,99000000)]

Now that, we have removed the outliers i.e cars which are less than 0 and more than 100,000,000 USD. Let's look at the number of rows in the dataset now. Initially, we had 50,000 rows.

In [22]:
autos.shape[0]

48578

We have about 48,578 rows now. Still a lot of rows, so our analysis will not change much.

## Exploring the date columns

There are 5 columns in the dataset that should represent date values. Some of these are created by the crawler and some came from the website itslef. They are:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as **string values**. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. 

Let's have a look at how the dates are formatted as strings.

In [23]:
autos[['date_crawled', 'last_seen', 'ad_created']].head()

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

As we can see above, the three columns are in the format of **YYYY-MM-DD HH:MM:SS**. If we just extract the first 10 characters from the string, we can extract the date. And then we could describe the data to calculate the distribution of values in percentages.

### **`date_crawled`** 

In [24]:
date_crawled = (autos['date_crawled']
               .str[:10]
               .value_counts(normalize=True, dropna=False)
               .sort_index())

date_crawled

2016-03-05    0.025320
2016-03-06    0.014039
2016-03-07    0.036004
2016-03-08    0.033328
2016-03-09    0.033101
2016-03-10    0.032175
2016-03-11    0.032566
2016-03-12    0.036930
2016-03-13    0.015666
2016-03-14    0.036539
2016-03-15    0.034275
2016-03-16    0.029602
2016-03-17    0.031640
2016-03-18    0.012907
2016-03-19    0.034769
2016-03-20    0.037877
2016-03-21    0.037404
2016-03-22    0.032998
2016-03-23    0.032216
2016-03-24    0.029334
2016-03-25    0.031599
2016-03-26    0.032196
2016-03-27    0.031084
2016-03-28    0.034851
2016-03-29    0.034131
2016-03-30    0.033678
2016-03-31    0.031846
2016-04-01    0.033678
2016-04-02    0.035469
2016-04-03    0.038598
2016-04-04    0.036519
2016-04-05    0.013092
2016-04-06    0.003170
2016-04-07    0.001400
Name: date_crawled, dtype: float64

* The dates crawled (information collected) are between the month of **March 5, 2016 and April 04, 2016** for a period of 34 days.
* The dates are almost evenly distributed, meaning the crawler collected the information regularly from the website.

### **`last_seen`**

In [25]:
last_seen = (autos['last_seen']
               .str[:10]
               .value_counts(normalize=True, dropna=False)
               .sort_index())

last_seen

2016-03-05    0.001070
2016-03-06    0.004323
2016-03-07    0.005393
2016-03-08    0.007431
2016-03-09    0.009613
2016-03-10    0.010663
2016-03-11    0.012372
2016-03-12    0.023797
2016-03-13    0.008893
2016-03-14    0.012598
2016-03-15    0.015871
2016-03-16    0.016448
2016-03-17    0.028079
2016-03-18    0.007349
2016-03-19    0.015830
2016-03-20    0.020647
2016-03-21    0.020647
2016-03-22    0.021368
2016-03-23    0.018527
2016-03-24    0.019762
2016-03-25    0.019206
2016-03-26    0.016798
2016-03-27    0.015645
2016-03-28    0.020874
2016-03-29    0.022356
2016-03-30    0.024764
2016-03-31    0.023797
2016-04-01    0.022788
2016-04-02    0.024929
2016-04-03    0.025197
2016-04-04    0.024476
2016-04-05    0.124768
2016-04-06    0.221808
2016-04-07    0.131912
Name: last_seen, dtype: float64

* The dates are similar to what we saw in the `date_crawled` column, for a period of 34 days. This makes sense because everytime an ad was crawled there will be a date when the ad was last seen.
* The distribution is almost similar to what we saw in the above `date_crawled` column.

### **`ad_created`**

In [26]:
ad_created = (autos['ad_created']
               .str[:10]
               .value_counts(normalize=True, dropna=False)
               .sort_index())

pd.set_option("display.max_rows", None) # doesnt collapse the value counts results

print(ad_created)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
2016-02-22    0.000021
2016-02-23    0.000082
2016-02-24    0.000041
2016-02-25    0.000062
2016-02-26    0.000041
2016-02-27    0.000124
2016-02-28    0.000206
2016-02-29    0.000165
2016-03-01    0.000103
2016-03-02    0.000103
2016-03-03    0.000865
2016-03-04    0.001482
2016-03-05    0.022891
2016-03-06 

* The advertisements created dates ranges from 11 June 2015 to 07 April 2016.
* We can see that, initially, the ads created were very low and there's no much activity on the webiste. But there's a sudden spike in the number of advertisements created from 05 March 2016 till the end of the range.

### **`registration_month`**

Since, the registration month is of numeric data type, there is no need of extracting anything.

In [27]:
month_registered = (autos['registration_month']
                   .value_counts(normalize=True)
                   .sort_index())

month_registered

0     0.092264
1     0.066306
2     0.060501
3     0.103030
4     0.083103
5     0.082980
6     0.087941
7     0.079419
8     0.064350
9     0.068550
10    0.073861
11    0.068200
12    0.069496
Name: registration_month, dtype: float64

* All months have an even distribution. 
* If we take the number 0 as the month of January, then 11 should be December, but there's a number with 12. So, we don't know what the numbers 0 and 12 mean in this column.
* So, we will drop this column, because there is no way of knowing what those numbers 0 and 12 mean.



In [28]:
autos = autos.drop(columns='registration_month')

### **`registration_year`**

Now let's have a look at the year of registration of ads on the website.

In [29]:
autos['registration_year'].describe()

count    48578.000000
mean      2004.753119
std         88.632571
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

* The data in this column seems to be very incorrect. 
* The minimum value is 1000 - before the cars were invented - and the crawler only collecte the information from the **year 2016**
* The maximum value is 9999 - way ahead into the future - which is beyond the year range in the `ad_created` and `date_crawled` column.
* Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.
* We will try to correct this column with our knowledge. Cars were first produced in the year 1885 and the maximum year we have in the `ad_created` column is 2016. So, we will take date only in the year range of 1900 - 2016.

In [30]:
autos = autos[autos['registration_year'].between(1900, 2016)]

Let's check the number of rows in our dataset now since we have considered only the cars which have a registration year between 1885 and 2016.

In [31]:
autos.shape # number of rows

(46692, 16)

# Data Analysis

So, now that we have cleaned our data to a certain extent by changing column names, converting their data types, dropping outliers and correcting incorrect data. We have made the data ready for our analysis. Let's find the answers to our questions.


### Exploring Price by Brand

In [32]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211278
bmw               0.110040
opel              0.107577
mercedes_benz     0.096441
audi              0.086546
ford              0.069926
renault           0.047139
peugeot           0.029834
fiat              0.025636
seat              0.018269
skoda             0.016405
nissan            0.015270
mazda             0.015185
smart             0.014157
citroen           0.014028
toyota            0.012700
hyundai           0.010023
sonstige_autos    0.009873
volvo             0.009145
mini              0.008760
mitsubishi        0.008224
honda             0.007839
kia               0.007068
alfa_romeo        0.006639
porsche           0.006125
suzuki            0.005932
chevrolet         0.005697
chrysler          0.003512
dacia             0.002634
daihatsu          0.002506
jeep              0.002270
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001563
daewoo            0.001499
trabant           0.001392
r

* The 6 most popular brands on the webiste are **volkswagen (21%), bmw (11%), opel (10%), mercedes_benz (9%), audi (8%), ford (6%).**
* They consistute almost **65%** of the entire cars listed on the website. So, we will focus on these cars to find our answers.

### Average price of the top 6 brands

Let's estimate the average price of the top 6 brands car mentioned above and see how they fare among the prices on the website. We will have the brands as a key and their average price as the values in the dictionary.

In [33]:
unique_brands = (autos['brand']
                 .value_counts()
                 .head(6)
                 .index) # index attribute which acceses the labels

average_price_per_brand = {} # prices in USD

for brand in unique_brands:
    mean_price = autos.loc[autos['brand'] == brand, 'price'].mean()
    average_price_per_brand[brand] = mean_price

average_price_per_brand

{'volkswagen': 6729.81956411556,
 'bmw': 8571.480147917478,
 'opel': 5432.479195699781,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 7456.547932618683}

* There's a distinct price gap in the top 6 brands.
> Audi, BMW and Mercedes Benz are more expensive, Ford and Opel are less expensive and Volkswagen is in between.
* After getting the average price of top 6 brands, we can see that **audi** has the highest average price for a used car in the German eBay website with a price of almost **9,336 USD.**
* The **second and third brand** with the highest average price is **mercedes benz** and **bmw** with the USD prices **8628** and **8751** respectively.
* The Germans prefer their own country manufactured cars when it comes to used cars.


Now that we have found out the mean price for each brands, we will check the **average mileage** for these cars and see if there's any visible link with the mean price.

### Average mileage of top 6 brands

In [34]:
average_mileage_per_brand = {}

for brand in unique_brands:
    mean_mileage = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    average_mileage_per_brand[brand] = mean_mileage
    
average_mileage_per_brand

{'volkswagen': 128713.6340598074,
 'bmw': 132575.9050214091,
 'opel': 129314.15488751742,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124243.49157733537}

In [35]:
# converting both dictionaries to a series objects
# using series constructor

average_price = pd.Series(average_price_per_brand)
average_mileage = pd.Series(average_mileage_per_brand)

# creating a dataframe using the dataframe constructor
df = pd.DataFrame(average_price, columns=['mean_price'])

# adding the other series in the new df
df['average_mileage'] = average_mileage

df # display the df

mean_price  average_mileage
volkswagen     6729.819564    128713.634060
bmw            8571.480148    132575.905021
opel           5432.479196    129314.154888
mercedes_benz  8628.450366    130788.363313
audi           9336.687454    129157.386785
ford           7456.547933    124243.491577

* When it comes to the mean mileage of the top 6 brands, the brand with highest mean mileage is **bmw** is **132,575 km**.
* The second and third are **mercedes benz** and **opel** with **130,788 km** and **129,314 km**, respectively.
* Coincidentally, two of these cars (bmw, mercedes benz) were also present in the top 3 cars with highest average price.
* Even though the cars have an higher average mileage value, they are expensive, even if it is a used only.

# Conclusion

* We saw the **top 6 brands** with the **highest average price.**
* Then we found out the **mean mileage of the top 6 brands.**
* We can say that, even though the car has a **higher mileage, those top 6 brands are very expensive in the used cars.**
* I believe, this is mainly due to the **brand value** which they posses in the industry and the car market. And also people are very **brand conscious**, willing to buy these cars even when they are **driven more and expensive, too.**

Some of the questions which can be answered further are:
1. How much cheaper are cars with damage than their non-damaged counterparts?
2. What are the most common brand/model combinations on the website?